galley/pkg/source/fs/source_test.go

```diff
@@ -186,117 +186,129 @@ spec:
 func TestNew(t *testing.T) {
 	dir := createTempDir(t)
 	defer deleteTempDir(t, dir)
-
-	_ = newOrFail(t, dir)
+	for _, watcherEnabled := range []bool{true, false} {
+		_ = newOrFail(t, dir, watcherEnabled)
+	}
 }
 
 func TestInvalidDirShouldSucceed(t *testing.T) {
-	s := newOrFail(t, "somebaddir")
+	for _, watcherEnabled := range []bool{true, false} {
+		s := newOrFail(t, "somebaddir", watcherEnabled)
```

Comment:
> I don't think it makes sense to do both paths for every test. e.g. for this test, the value of watcherEnabled won't matter at all.

@sushicw IMO, wherever we are creating a new source i.e calling `newOrFail()`, should be tested with both the values of  `watcherEnabled`. Why? to ensure that behaviour is consistent and enabling fsnotify is not changing the expected outcome. For example in this test case, this is guarding the logic/behaviour in case if some in future makes a change to alter behaviour in case of invalidDir. If we dont test with both values, we are leaving one flow uncovered.

Reply:
I agree that both flows should be covered. However, when more than one test invokes is covering the same part of the flow, it becomes redundant to have them all test that part in detail. This also breaks the principle that ideally each test should test exactly one thing and makes the tests less specific and more brittle. 

Put another way: if the behavior does change unexpectedly, we don't need *every* test to fail, and having every test fail actually might even make it harder to debug. And if a behavior change wouldn't impact the result of the test either way, then it doesn't belong in that test. 

e.g. the value of enableFsnotify doesn't impact the New function at all, so there's no reason for TestNew to check both values. 

In addition, what makes me nervous about the for loops here is that when the test fails, it will be difficult to tell which iteration it was on.

The approach I'd recommend most would be to create a brand-new test that specifically covers the new behavior. I'd start by making a copy of TestBuiltinResource, since that's the simplest testy that invokes the add/update/delete behavior we need to cover. Modify it to use the watcher instead of manually signalling. You should also probably use ```g.Eventually``` instead of ```g.Expect``` to verify that the update was received, to avoid test race conditions. 

I believe copying or modifying the other tests would be redundant.